atef/config.py

```diff
@@ -140,22 +172,162 @@ class PreparedComparison:
+
+        Parameters
+        ----------
+        config : PVConfiguration or DeviceConfiguration
+            The configuration.
+        client : happi.Client, optional
+            A happi Client instance.
+        cache : DataCache, optional
+            The data cache to use for this and other similar comparisons.
+
+        Yields
+        ------
+        item : PreparedComparisonException or PreparedComparison
+            If an error occurs during preparation, a
+            PreparedComparisonException will be yielded in place of the
+            PreparedComparison.
+        """
+        if cache is None:
+            cache = DataCache()
+
+        if isinstance(config, PVConfiguration):
+            yield from PreparedSignalComparison._from_pv_config(config, cache=cache)
+        elif isinstance(config, DeviceConfiguration):
+            if client is None:
+                client = happi.Client.from_config()
+            for dev_name in config.devices:
+                try:
+                    device = util.get_happi_device_by_name(dev_name, client=client)
+                except Exception as ex:
+                    yield PreparedComparisonException(
+                        exception=ex,
+                        comparison=None,  # TODO
+                        name=config.name or config.description,
+                        identifier=dev_name,
+                        path=[
+                            config,
+                            dev_name,
+                        ],
+                    )
+                else:
+                    yield from PreparedSignalComparison._from_device_config(
+                        config=config,
+                        device=device,
+                        cache=cache,
+                    )
+        elif isinstance(config, ToolConfiguration):
+            yield from PreparedToolComparison._from_tool_config(config, cache=cache)
+        else:
+            raise NotImplementedError(f"Configuration type unsupported: {type(config)}")
+
+
+@dataclass
+class PreparedSignalComparison(PreparedComparison):
+    """
+    A unified representation of comparisons for device signals and standalone
+    PVs.
+
+    Each PreparedSignalComparison has a single leaf in the configuration tree,
+    comprised of:
+    * A configuration
+    * The signal specification.  This is comprised of the configuration and
+        "IdentifierAndComparison"
+        - DeviceConfiguration: Device and attribute (the "identifier")
+        - PVConfiguration: PV name (the "identifier")
+    * A comparison to run
+        - Including data reduction settings
+    """
+    #: The device the comparison applies to, if applicable.
+    device: Optional[ophyd.Device] = None
+    #: The signal the comparison is to be run on.
+    signal: Optional[ophyd.Signal] = None
+    #: The value from the signal the comparison is to be run on.
+    data: Optional[Any] = None
+
+    async def get_data_async(self) -> Any:
```

Comment:
Note to self: this is where we'd put the bit that gets the dynamic comparison's values, if any exist.